tests/Collection/CrudSpecFunctionalTest.php

```diff
@@ -212,6 +233,42 @@ private function executeOutcome(array $operation, array $outcome, $actualResult)
         }
     }
 
+    /**
+     * Extracts a result from an exception.
+     *
+     * Errors for bulkWrite and insertMany operations may still report a write
+     * result. This method will attempt to extract such a result so that it can
+     * be used in executeAssertResult().
+     *
+     * If no result can be extracted, null will be returned.
+     *
+     * @param array            $operation
+     * @param RuntimeException $exception
+     * @return mixed
+     */
+    private function extractResultFromException(array $operation, array $outcome, RuntimeException $exception)
+    {
+        switch ($operation['name']) {
+            case 'bulkWrite':
+                $insertedIds = isset($outcome['result']['insertedIds']) ? $outcome['result']['insertedIds'] : [];
+
+                if ($exception instanceof BulkWriteException) {
+                    return new BulkWriteResult($exception->getWriteResult(), $insertedIds);
+                }
+                break;
+
+            case 'insertMany':
+                $insertedIds = isset($outcome['result']['insertedIds']) ? $outcome['result']['insertedIds'] : [];
```

Comment:
Per https://github.com/mongodb/specifications/pull/399#discussion_r222099362, the spec tests will not actually contain `insertedIds` in the expected result document; however, I would like our test runner to respect `insertedIds` if it's present.

In practice, a spec test with `insertedIds` would cause our driver to fail as our BulkWrite and InsertMany operations have no way of injecting inserted IDs into the thrown BulkWriteException. Doing so would require a custom PHPLIB exception class that contained a PHPLIB write result (which does include inserted IDs).